integration/networking/port_mapping_linux_test.go

```diff
@@ -0,0 +1,313 @@
+	assert.Check(t, is.Contains(res.Stderr.String(), "404 Not Found"))
+}
+
+// Check that a container on an IPv4-only network can have a port mapping
+// from a specific IPv6 host address (using docker-proxy).
+// Regression test for https://github.com/moby/moby/issues/48067 (which
+// is about incorrectly reporting this as invalid config).
+func TestProxy4To6(t *testing.T) {
+	skip.If(t, testEnv.IsRootless)
+
+	ctx := setupTest(t)
+	d := daemon.New(t)
+	d.StartWithBusybox(ctx, t)
+	defer d.Stop(t)
+
+	c := d.NewClientT(t)
+	defer c.Close()
+
+	const netName = "ipv4net"
+	network.CreateNoError(ctx, t, c, netName)
+
+	serverId := container.Run(ctx, t, c,
+		container.WithNetworkMode(netName),
+		container.WithExposedPorts("80"),
+		container.WithPortMap(nat.PortMap{"80": {{HostIP: "::1"}}}),
+		container.WithCmd("httpd", "-f"),
+	)
+	defer c.ContainerRemove(ctx, serverId, containertypes.RemoveOptions{Force: true})
+
+	inspect := container.Inspect(ctx, t, c, serverId)
+	hostPort := inspect.NetworkSettings.Ports["80/tcp"][0].HostPort
+
+	resp, err := http.Get("http://[::1]:" + hostPort)
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal(resp.StatusCode, 404))
+}
+
+// TestAccessPublishedPortFromLocalHost checks whether published ports are
+// accessible from the host netns.
+func TestAccessPublishedPortFromLocalHost(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
+	skip.If(t, testEnv.IsRootless())
+
+	ctx := setupTest(t)
+
+	assert.NilError(t, exec.Command("sysctl", "net.ipv6.conf.all.disable_ipv6=0").Run())
+	out, err := exec.Command("ip", "addr", "add", "fd85:1234:4321:cafe::1/64", "dev", "eth0").CombinedOutput()
+	if err != nil && !strings.Contains(string(out), "address already assigned") {
+		assert.NilError(t, err, string(out))
+	}
+
+	testcases := []struct {
+		name        string
+		daemonFlag  string
+		clientIface string
+		ipv6        bool
+	}{
+		{
+			name:        "userland-proxy=true/IPv4/loopback address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "lo",
+		},
+		{
+			name:        "userland-proxy=true/IPv4/local IP address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "eth0",
+		},
+		{
+			name:        "userland-proxy=false/IPv4/loopback address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "lo",
+		},
+		{
+			name:        "userland-proxy=false/IPv4/local IP address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "eth0",
+		},
+		{
+			name:        "userland-proxy=true/IPv6/local IP address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "eth0",
+			ipv6:        true,
+		},
+		{
+			name:        "userland-proxy=false/IPv6/local IP address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "eth0",
+			ipv6:        true,
+		},
+	}
+
+	for tcID, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := testutil.StartSpan(ctx, t)
+
+			clientAddr, err := getIfaceAddress(t, tc.clientIface, tc.ipv6)
```

Comment:
`clientAddr` -> `hostAddr` ... as it's a server address, on the host? (Ditto for `clientIface` -> `hostIface`.)